File: dialog-hub/src/main/java/com/jnewstrom/screenswitcher/dialoghub/ScopedDialogFactory.java
package com.jnewstrom.screenswitcher.dialoghub;

import android.app.Dialog;
import android.content.Context;

import com.jaynewstrom.concrete.Concrete;
import com.jaynewstrom.concrete.ConcreteWall;

public abstract class ScopedDialogFactory<C> implements DialogFactory {
    private final ConcreteWall<C> concreteWall;

    public ScopedDialogFactory(Context scopedContext) {
        concreteWall = Concrete.findWall(scopedContext);
    }

    @Override public final Dialog createDialog(Context context) {
        Context scopedContext = concreteWall.createContext(context);
        return createDialog(scopedContext, concreteWall.getComponent());
    }

    protected abstract Dialog createDialog(Context context, C component);
}

File: dialog-manager/src/main/java/com/jaynewstrom/screenswitcher/dialogmanager/ScopedDialogFactory.kt
package com.jaynewstrom.screenswitcher.dialogmanager

import android.app.Dialog
import android.content.Context

import com.jaynewstrom.concrete.Concrete
import com.jaynewstrom.concrete.ConcreteWall

abstract class ScopedDialogFactory<in C>(scopedContext: Context) : DialogFactory {
    private val concreteWall: ConcreteWall<C> = Concrete.findWall(scopedContext)

    override fun createDialog(context: Context): Dialog {
        val scopedContext = concreteWall.createContext(context)
        return createDialog(scopedContext, concreteWall.component)
    }

    protected abstract fun createDialog(context: Context, component: C): Dialog
}
